File: grid_generator/src/index_helpers.f90
! This source file is part of the Geophysical Fluids Modeling Framework (GAME), which is released under the MIT license.
! Github repository: https://github.com/OpenNWP/GAME

module index_helpers
  
  ! This module contains helper functions concerned with simple algebraic operations on vectors.

  use iso_c_binding
  
  implicit none
  
  private
  
  public :: find_min_index
  public :: find_max_index
  public :: find_min_index_exclude
  public :: in_bool_calculator
  
  contains

  function find_min_index(vector,vector_length) &
  bind(c,name = "find_min_index")
  
    ! This function returns the index where a vector has its minimum.
    
    real(c_double), intent(in) :: vector(vector_length)
    integer, intent(in)        :: vector_length
    integer                    :: find_min_index
    
    ! local variables
    integer        :: ji
    real(c_double) :: current_min
    
    find_min_index = 1
    current_min = vector(1)
    
    do ji=2,vector_length
      if (vector(ji)<current_min) then
        current_min = vector(ji) 
        find_min_index = ji
      endif
    enddo
    
    find_min_index = find_min_index - 1
    
  end function find_min_index
  
  function find_min_index_exclude(vector,vector_length,exclude_indices_vector,exclude_indices_vector_length) &
  bind(c,name = "find_min_index_exclude")
  
    ! This function finds the index where a vector has its minimum, excluding the elements of another vector.
    
    real(c_double), intent(in) :: vector(vector_length)
    integer, intent(in)        :: vector_length
    integer, intent(in)        :: exclude_indices_vector(exclude_indices_vector_length)
    integer, intent(in)        :: exclude_indices_vector_length
    integer                    :: find_min_index_exclude
    
    ! local variables
    integer        :: ji
    real(c_double) :: current_min
    
    current_min = maxval(vector) + 1.0
    find_min_index_exclude = 0
    
    do ji=1,vector_length
      if (vector(ji)<current_min) then
        if (in_bool_calculator(ji,exclude_indices_vector,exclude_indices_vector_length)==0) then
          current_min = vector(ji)
          find_min_index_exclude = ji
        endif
      endif
    enddo
    
    find_min_index_exclude = find_min_index_exclude - 1
    
    if (find_min_index_exclude==-1) then
      write(*,*) "Function find_min_index_exclude failed."
      call exit(1)
    endif
    
  end function find_min_index_exclude

  function find_max_index(vector,vector_length) &
  bind(c,name = "find_max_index")
  
    ! This function returns the index where a vector has its maximum.
    
    real(c_double), intent(in) :: vector(vector_length)
    integer, intent(in)        :: vector_length
    integer                    :: find_max_index
    
    ! local variables
    integer        :: ji
    real(c_double) :: current_min
    
    find_max_index = 1
    current_min = vector(1)
    
    do ji=2,vector_length
      if (vector(ji)>current_min) then
        current_min = vector(ji) 
        find_max_index = ji
      endif
    enddo
    
    find_max_index = find_max_index - 1
    
  end function find_max_index


  function in_bool_calculator(value,vector,vector_length) &
  bind(c,name = "in_bool_calculator")

    ! This function checks if a vector of integers contains a certain value.
    
    integer, intent(in) :: value
    integer, intent(in) :: vector(vector_length)
    integer, intent(in) :: vector_length
    integer             :: in_bool_calculator
    
    ! local variables
    integer :: ji
    
    in_bool_calculator = 0

    do ji=1,vector_length
      if(vector(ji)==value) then
        in_bool_calculator = 1
        exit
      endif
    enddo

  end function in_bool_calculator

end module index_helpers

File: grid_generator/src/index_helpers.c
/*
This source file is part of the Geophysical Fluids Modeling Framework (GAME), which is released under the MIT license.
Github repository: https://github.com/OpenNWP/GAME
*/

/*
This file contains helper functions concerned with simple algebraic operations on vectors.
*/

#include <math.h>
#include "grid_generator.h"

int find_min_index_exclude(double vector[], int vector_length, int exclude_indices_vector[], int exclude_indices_vector_length)
{
	/*
	This function finds the index where a vector has its minimum, excluding the elements of another vector.
	*/
    int result = 0;
    double current_min = vector[0];
    for (int i = 1; i < vector_length; ++i)
    {
        if (vector[i] < current_min)
        {
        	if (in_bool_calculator(&i, exclude_indices_vector, &exclude_indices_vector_length) == 0)
        	{
            	current_min = vector[i];
            	result = i;
        	}
        }
    }
    return result;
}

double double_sum_gen(double vector[], int vector_length, int first_index, int second_index)
{
	/*
	This function calculates the sum of all elements of a vector of doubles between two indices.
	*/
	double result = 0.0;
	if (first_index <= second_index)
	{
		for (int i = first_index ; i <= second_index; ++i)
		{
			result += vector[i];
		}
	}
	else
	{
		for (int i = first_index; i < vector_length; ++i)
		{
			result += vector[i];
		}
		for (int i = 0; i <= second_index; ++i)
		{
			result += vector[i];
		}	
	}
	return result;
}









